File: src/main/java/engineer/carrot/warren/warren/irc/handlers/core/PingHandler.java
package engineer.carrot.warren.warren.irc.handlers.core;

import engineer.carrot.warren.warren.event.ServerPingEvent;
import engineer.carrot.warren.warren.irc.handlers.MessageHandler;
import engineer.carrot.warren.warren.irc.messages.core.PingMessage;
import engineer.carrot.warren.warren.irc.messages.core.PongMessage;

public class PingHandler extends MessageHandler<PingMessage> {
    @Override
    public void handleMessage(PingMessage message) {
        this.outgoingQueue.addMessageToQueue(new PongMessage(message.pingToken));

        this.eventSink.postEvent(new ServerPingEvent(message.pingToken, message.pingToken));
    }
}

File: src/main/kotlin/engineer/carrot/warren/warren/handler/PingHandler.kt
package engineer.carrot.warren.warren.handler

import engineer.carrot.warren.kale.IKaleHandler
import engineer.carrot.warren.kale.irc.message.rfc1459.PingMessage
import engineer.carrot.warren.kale.irc.message.rfc1459.PongMessage
import engineer.carrot.warren.warren.IMessageSink
import engineer.carrot.warren.warren.loggerFor
import engineer.carrot.warren.warren.state.ConnectionState

class PingHandler(val sink: IMessageSink, val connectionState: ConnectionState) : IKaleHandler<PingMessage> {

    private val LOGGER = loggerFor<PingHandler>()

    override val messageType = PingMessage::class.java

    override fun handle(message: PingMessage, tags: Map<String, String?>) {
        LOGGER.debug("handling ping with token ${message.token}")

        sink.write(PongMessage(token = message.token))

        connectionState.lastPingOrPong = System.currentTimeMillis()
    }

}